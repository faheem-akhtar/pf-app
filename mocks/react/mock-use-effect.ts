type OnUnmount = () => void;

const original = global.React.useEffect;

const unmountCallbacks: Function[] = [];

export const reactMockUseEffect = (): { unmountAll: OnUnmount } => {
  jest.spyOn(global.React, 'useEffect').mockImplementation((f) => {
    const onUnmount = f();
    if (onUnmount) {
      unmountCallbacks.push(onUnmount);
    }
  });

  return {
    unmountAll: (): void => {
      unmountCallbacks.forEach((cb) => cb());
      unmountCallbacks.length = 0;
    },
  };
};

// eslint-disable-next-line pf-rules/export-name-validation
export const recoverReactUseEffect = (): void => {
  unmountCallbacks.length = 0;
  global.React.useEffect = original;
};
