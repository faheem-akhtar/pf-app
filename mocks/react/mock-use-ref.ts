const original = global.React.useRef;

const unmountCallbacks: Function[] = [];

export const mockReactUseRef = <T>(initialValue?: T): React.MutableRefObject<T> => {
  const mock = { current: initialValue } as React.MutableRefObject<T>;
  const mockImpl: any = (initial: T) => {
    if (initialValue === undefined) {
      mock.current = initial;
    }
    return mock;
  };

  jest.spyOn(global.React, 'useRef').mockImplementation(mockImpl);
  return mock;
};

// eslint-disable-next-line pf-rules/export-name-validation
export const recoverReactUseRef = (): void => {
  unmountCallbacks.length = 0;
  global.React.useRef = original;
};
