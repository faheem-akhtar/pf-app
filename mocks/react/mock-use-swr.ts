import { windowMockFetch } from 'mocks/window/mock-fetch';

import { AnyValueType } from 'types/any/value.type';
import { ApiSwrResultType } from 'api/swr-result-type';
import { SWRResponse } from 'swr';

let resultData: Partial<ApiSwrResultType<AnyValueType>>;
let isValidating = false;

export const reactMockUseSwr = (
  swrResultData: Partial<ApiSwrResultType<AnyValueType>>,
  swrIsValidating: boolean = false
): void => {
  resultData = swrResultData;
  isValidating = swrIsValidating;
};

// eslint-disable-next-line pf-rules/export-name-validation
export const setupSwrMock = (): void => {
  jest.mock('swr', () => (shouldFetch: boolean, fetcher: Function): SWRResponse<AnyValueType, AnyValueType> => {
    const baseProps = {
      data: null,
      isValidating: true,
      revalidate: (): Promise<boolean> => Promise.resolve(true),
      mutate: (): Promise<null> => Promise.resolve(null),
    };
    if (shouldFetch) {
      windowMockFetch();
      fetcher();
      return {
        ...baseProps,
        data: resultData,
        isValidating,
      };
    }
    return baseProps;
  });
};
