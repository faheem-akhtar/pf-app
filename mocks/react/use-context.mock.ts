import { AnyValueType } from 'types/any/value.type';

const original = global.React.useContext;

export const mockReactUseContext = (...contextsValues: AnyValueType[]): void => {
  let callNumber = 0;
  jest.spyOn(global.React, 'useContext').mockImplementation(() => {
    if (callNumber === contextsValues.length) {
      throw new Error(`useContext was called for ${callNumber + 1} time, but mock context value was not specified.`);
    }
    return contextsValues[callNumber++];
  });
};

// eslint-disable-next-line @propertyfinder/rules/export-name-validation
export const recoverReactUseContext = (): void => {
  global.React.useContext = original;
};
