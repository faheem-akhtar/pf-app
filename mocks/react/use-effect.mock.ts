type OnUnmount = () => void;

const original = global.React.useEffect;

const unmountCallbacks: Function[] = [];

type Props = {
  callImmediately: boolean;
};

export const mockReactUseEffect = (props?: Props): { trigger: () => void; unmountAll: OnUnmount } => {
  const { callImmediately = true } = props || {};

  let trigger: () => void;

  jest.spyOn(global.React, 'useEffect').mockImplementation((f) => {
    trigger = (): void => {
      const onUnmount = f();
      if (onUnmount) {
        unmountCallbacks.push(onUnmount);
      }
    };

    if (callImmediately) {
      trigger();
    }
  });

  return {
    trigger: (): void => trigger && trigger(),
    unmountAll: (): void => {
      unmountCallbacks.forEach((cb) => cb());
      unmountCallbacks.length = 0;
    },
  };
};

// eslint-disable-next-line pf-rules/export-name-validation
export const recoverReactUseEffect = (): void => {
  unmountCallbacks.length = 0;
  global.React.useEffect = original;
};
