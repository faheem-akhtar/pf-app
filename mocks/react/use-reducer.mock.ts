const original = global.React.useReducer;

export const mockReactUseReducer = <State>(
  state: State
): {
  useReducerMock: jest.Mock<State>;
  dispatchMock: jest.Mock;
} => {
  const useReducerMock = jest.fn();
  const dispatchMock = jest.fn();
  useReducerMock.mockReturnValue([state, dispatchMock]);
  jest.spyOn(global.React, 'useReducer').mockImplementation(useReducerMock);

  return {
    useReducerMock,
    dispatchMock,
  };
};

// eslint-disable-next-line @propertyfinder/rules/export-name-validation
export const recoverReactUseReducer = (): void => {
  global.React.useReducer = original;
};
