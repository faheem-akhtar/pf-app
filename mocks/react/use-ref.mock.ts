const original = global.React.useRef;

const unmountCallbacks: Function[] = [];

export const mockReactUseRef = <T>(initialValue?: T): React.MutableRefObject<T> => {
  const mock = { current: initialValue } as React.MutableRefObject<T>;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const mockImpl: any = (initial: T) => {
    if (initialValue === undefined && mock.current === undefined) {
      mock.current = initial;
    }
    return mock;
  };

  jest.spyOn(global.React, 'useRef').mockImplementation(mockImpl);
  return mock;
};

// eslint-disable-next-line @propertyfinder/rules/export-name-validation
export const recoverReactUseRef = (): void => {
  unmountCallbacks.length = 0;
  global.React.useRef = original;
};
