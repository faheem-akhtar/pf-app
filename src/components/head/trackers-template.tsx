import { configAnalyticsGtmId } from 'config/analytics/gtm-id';
import { configCommon } from 'config/common';
import { configTealiumEnabled } from 'config/tealium/enabled';
import { tealiumProfileId } from 'config/tealium/profile-id';
import { helpersIsDevelopment } from 'helpers/is-development';
import { AnalyticsPageNamesEnum } from 'services/analytics/page-names.enum';
import { AnalyticsTealiumService } from 'services/analytics/tealium.service';
import { TealiumEventEnum } from 'services/tealium/event.enum';
import { TealiumPageTypeEnum } from 'services/tealium/page-type.enum';

const __html = `
function urlQueryGetParameterByName(name, url = window.location.href) {
  name = name.replace(/[[]]/g, '\\$&');
  var regex = new RegExp('[?&]' + name + '(=([^&#]*)|&|#|$)'),
      results = regex.exec(url);
  if (!results) return null;
  if (!results[2]) return '';
  return decodeURIComponent(results[2].replace(/\\+/g, ' '));
}

window.dataLayer = [];

if (!urlQueryGetParameterByName('no-gtm')) {
  // GTM
  (function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
  new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
  j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
  '//www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);
  })(window,document,'script','dataLayer', '${configAnalyticsGtmId}');
}

if (!urlQueryGetParameterByName('no-analytics')) {
  if (!urlQueryGetParameterByName('no-tealium') && ${configTealiumEnabled}) {
    // https://docs.tealium.com/platforms/javascript/settings
    window.utag_cfg_ovrd = {
      readywait: true
    }

    // TODO-FE[CX-186] setup tealium variables from env
    // Default Tealium event data
    window.tealium = {
      "page_country"        : "${configCommon.countryCode}",                           //Required
      "page_currency_code"  : "${configCommon.currencyCode}",             //Required
      "page_lang"           : "${configCommon.language.current}",                        //Required
      "page_category"       : "${AnalyticsPageNamesEnum.ancillaryPage}",  //Required
      "page_type"           : "${TealiumPageTypeEnum.other}",                //Required
      "tealium_event"       : "${TealiumEventEnum.default}",      //Required
    };

    // Tealium
    (function(a,b,c,d){
      a='https://tags.tiqcdn.com/utag/propertyfinder/${tealiumProfileId}/prod/utag.js';
      b=document;c='script';d=b.createElement(c);d.src=a;d.type='text/java'+c;d.async=true;
      d.onload=function(){ ${AnalyticsTealiumService.callStalledEvents()} };
      a=b.getElementsByTagName(c)[0];a.parentNode.insertBefore(d,a);
      })();
  }

  // Snowplow
  if (!urlQueryGetParameterByName('no-snowplow')) {
    if (window.localStorage) {
      ;(function(p,l,o,w,i,n,g){if(!p[i]){p.GlobalSnowplowNamespace=p.GlobalSnowplowNamespace||[];
          p.GlobalSnowplowNamespace.push(i);p[i]=function(){(p[i].q=p[i].q||[]).push(arguments)
          };p[i].q=p[i].q||[];n=l.createElement(o);g=l.getElementsByTagName(o)[0];n.async=1;
          n.src=w;g.parentNode.insertBefore(n,g)}}(window,document,"script","/en/static/sp.js","snowplow"));

      window.snowplow(
        'newTracker',
        'pf-website',
        '${process.env.NEXT_PUBLIC_SNOWPLOW_HOST}',
        {
          appId: 'pf-mobilesite',
          platform: 'web',
          discoverRootDomain: true,
          bufferSize: ${helpersIsDevelopment ? 1 : 5},
          respectDoNotTrack: false,
          forceSecureTracker: true,
          post: true,
          contexts: {
              webPage: true,
              performanceTiming: true,
              gaCookies: true,
              geolocation: false
          }
        }
      );

      window.snowplow('enableActivityTracking', 30, 10);

      window.snowplow('enableLinkClickTracking');

      window.snowplow('crossDomainLinker', function(linkElement) {
        return /^https:\\/\\/www\\.mortgagefinder\\.ae/.test(linkElement.href);
      });
    }
  }
}
`;

export const HeadTrackersTemplate = (): JSX.Element => (
  // eslint-disable-next-line react/no-danger
  <script dangerouslySetInnerHTML={{ __html }} />
);
