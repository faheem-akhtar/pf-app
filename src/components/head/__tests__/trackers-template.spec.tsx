import { render } from '@testing-library/react';

import { HeadTrackersTemplate } from '../trackers-template';

describe('HeadTrackersTemplate', () => {
  it('should render without error', () => {
    const { container } = render(<HeadTrackersTemplate snowplowHost={'c.propertyfinderdata.com'} />);

    expect(container.innerHTML).toMatchSnapshot();
  });
});
