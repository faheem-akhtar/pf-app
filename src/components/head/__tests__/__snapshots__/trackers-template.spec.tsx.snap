// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeadTrackersTemplate should render without error 1`] = `
"<script>
function urlQueryGetParameterByName(name, url = window.location.href) {
  name = name.replace(/[[]]/g, '\\\\$&');
  var regex = new RegExp('[?&]' + name + '(=([^&#]*)|&|#|$)'),
      results = regex.exec(url);
  if (!results) return null;
  if (!results[2]) return '';
  return decodeURIComponent(results[2].replace(/\\\\+/g, ' '));
}

window.dataLayer = [];

if (!urlQueryGetParameterByName('no-gtm')) {
  // GTM
  (function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
  new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
  j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
  '//www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);
  })(window,document,'script','dataLayer', 'GTM-KNNMB4');
}

if (!urlQueryGetParameterByName('no-analytics')) {
  // TODO-FE[CX-186] remove false
  if (false && !urlQueryGetParameterByName('no-tealium')) {
    // TODO-FE[CX-186] setup tealium variables from env
    // Default Tealium event data
    window.tealium = {
      \\"page_country\\"        : \\"ae\\",                           //Required
      \\"page_currency_code\\"  : \\"AED\\",             //Required
      \\"page_lang\\"           : \\"en\\",                        //Required
      \\"page_category\\"       : \\"Ancillary Page\\",  //Required
      \\"page_type\\"           : \\"other\\",                //Required
      \\"tealium_event\\"       : \\"view\\",      //Required
    };

    // Tealium
    (function(a,b,c,d){
    a='https://tags.tiqcdn.com/utag/propertyfinder/uae/prod/utag.js';
    b=document;c='script';d=b.createElement(c);d.src=a;d.type='text/java'+c;d.async=true;
    a=b.getElementsByTagName(c)[0];a.parentNode.insertBefore(d,a);
    })();
  }

  // Snowplow
  if (!urlQueryGetParameterByName('no-snowplow')) {
    if (window.localStorage) {
      ;(function(p,l,o,w,i,n,g){if(!p[i]){p.GlobalSnowplowNamespace=p.GlobalSnowplowNamespace||[];
          p.GlobalSnowplowNamespace.push(i);p[i]=function(){(p[i].q=p[i].q||[]).push(arguments)
          };p[i].q=p[i].q||[];n=l.createElement(o);g=l.getElementsByTagName(o)[0];n.async=1;
          n.src=w;g.parentNode.insertBefore(n,g)}}(window,document,\\"script\\",\\"/static/sp.js\\",\\"snowplow\\"));

      window.snowplow(
        'newTracker',
        'pf-website',
        'c.propertyfinderdata.com',
        {
          appId: 'pf-mobilesite',
          platform: 'web',
          discoverRootDomain: true,
          bufferSize: 5,
          respectDoNotTrack: false,
          forceSecureTracker: true,
          post: true,
          contexts: {
              webPage: true,
              performanceTiming: true,
              gaCookies: true,
              geolocation: false
          }
        }
      );

      window.snowplow('enableActivityTracking', 30, 10);

      window.snowplow('enableLinkClickTracking');

      window.snowplow('crossDomainLinker', function(linkElement) {
        return /^https:\\\\/\\\\/www\\\\.mortgagefinder\\\\.ae/.test(linkElement.href);
      });
    }
  }
}
</script>"
`;
